### PR TITLE
[transformer_engine executor] Update args to `_Linear.forward` and `Context`

### PR DESCRIPTION
**Pull request by crcrpar**
Switch to a dict from a sequence of arguments for a bit better solidity, also some optimistic patch hoping TransformerEngine executor a bit more nicely supports nvidia/transformerengine repo.

### CODE REVIEW

**Comment by kshitij12345** [comment]
I think the patch should be for version above 1.5. There is an issue regarding the breakage of this integration : https://github.com/Lightning-AI/lightning-thunder/issues/117. Also, we are talking to the TE team to get more stable API so that we don't have such breakages.

This patch is failing for 1.5
```
TypeError: _Linear.forward() got an unexpected keyword argument 'skip_fp8_weight_update'
```

**Comment by kshitij12345** [comment]
Following patch worked for current [stable](https://github.com/NVIDIA/TransformerEngine/tree/stable) (v1.5) and v1.6 [branch](https://github.com/NVIDIA/TransformerEngine/tree/release_v1.6)

```patch
diff --git a/thunder/executors/transformer_engineex.py b/thunder/executors/transformer_engineex.py
index 10af787..b922a06 100644
--- a/thunder/executors/transformer_engineex.py
+++ b/thunder/executors/transformer_engineex.py
@@ -35,7 +35,7 @@ TE_AVAILABLE: bool = package_available("transformer_engine")
 # between version 1.2 and 1.3.
 # Hence, we have these guards based on version.
 TE_VERSION_1_3_PLUS: bool = False
-TE_VERSION_1_5_PLUS: bool = False
+TE_VERSION_1_6_PLUS: bool = False
 
 te: None | Any = None
 if TE_AVAILABLE:
@@ -49,7 +49,7 @@ if TE_AVAILABLE:
         TE_AVAILABLE = False
 
     TE_VERSION_1_3_PLUS = LooseVersion(version("transformer_engine")) >= LooseVersion("1.3")
-    TE_VERSION_1_5_PLUS = LooseVersion(version("transformer_engine")) > LooseVersion("1.5")
+    TE_VERSION_1_6_PLUS = LooseVersion(version("transformer_engine")) >= LooseVersion("1.6")
 if not TE_AVAILABLE:
     TransformerEngineBaseModule = object
 
@@ -135,8 +135,11 @@ class Context:
 
         if TE_VERSION_1_3_PLUS:
             ctx_dict["cpu_offloading"] = self.cpu_offloading
-        if TE_VERSION_1_5_PLUS:
+        if TE_VERSION_1_6_PLUS:
             ctx_dict["primary_weights_in_fp8"] = self.primary_weights_in_fp8
+            ctx_dict["is_input_fp8"] = self.is_input_fp8
+            ctx_dict["reduce_and_update_bwd_fp8_tensors"] = self.reduce_and_update_bwd_fp8_tensors
+            ctx_dict["ub_overlap_ag"] = self.ub_overlap_ag
         else:
             ctx_dict.update(
                 {
@@ -166,8 +169,11 @@ class Context:
         ctx.requires_dgrad = d["requires_dgrad"]
         if TE_VERSION_1_3_PLUS:
             ctx.cpu_offloading = d["cpu_offloading"]
-        if TE_VERSION_1_5_PLUS:
+        if TE_VERSION_1_6_PLUS:
             ctx.primary_weights_in_fp8 = d["primary_weights_in_fp8"]
+            ctx.is_input_fp8 = d["is_input_fp8"]
+            ctx.reduce_and_update_bwd_fp8_tensors = d["reduce_and_update_bwd_fp8_tensors"]
+            ctx.ub_overlap_ag = d["ub_overlap_ag"]
         else:
             ctx.ub_split_ag = d["ub_split_ag"]
             ctx.ub_atomic_gemm_ag = d["ub_atomic_gemm_ag"]
@@ -260,7 +266,7 @@ class TELinear(TransformerEngineBaseModule):
             }
             if TE_VERSION_1_3_PLUS:
                 kwargs["cpu_offloading"] = CPUOffloadEnabled
-            if TE_VERSION_1_5_PLUS:
+            if TE_VERSION_1_6_PLUS:
                 kwargs.update(
                     {
                         # ref: https://github.com/NVIDIA/TransformerEngine/blame/7c1828f80edc1405d4ef1a7780c9e0046beab
5c7/transformer_engine/pytorch/module/linear.py#L70
diff --git a/thunder/tests/test_transformer_engine_executor.py b/thunder/tests/test_transformer_engine_executor.py
index e8abcc7..9d59de8 100644
--- a/thunder/tests/test_transformer_engine_executor.py
+++ b/thunder/tests/test_transformer_engine_executor.py
@@ -174,10 +174,9 @@ def test_te_with_autocast():
     x = torch.randn(16, 16, device=device, requires_grad=True)
     w = torch.randn(16, 16, device=device, requires_grad=True)
 
-    cfunc = thunder.compile(
+    cfunc = thunder.jit(
         thunder.core.transforms.autocast(foo, dtype=thunder.dtypes.bfloat16),
-        executors_list=[transformer_engine_ex],
-        disable_preprocessing=True,
+        executors=[transformer_engine_ex,],
     )
     cfunc(x, w)
 
```

**Comment by kshitij12345** [comment]
Actually, I think we should simplify `Context` as following, so that we don't need to care about exact information being saved for backward. I have verified this on v1.5 (stable), v1.6 (next release) and v1.7 (main) and it passes the tests on all three.

cc: @IvanYashchuk 

```python
class Context:
    def __init__(self):
        self.saved_tensors = ()

    def save_for_backward(self, *tensors):
        self.saved_tensors = tensors

    def to_dict(self):
        return self.__dict__

    @staticmethod
    def from_dict(d):
        ctx = Context()
        for key, value in d.items():
            setattr(ctx, key, value)
        return ctx

```

**Comment by IvanYashchuk** [comment]
Let's merge this pull request once @kshitij12345 is happy with the changes and in a follow-up one try using the suggestion from https://github.com/Lightning-AI/lightning-thunder/pull/208#issuecomment-2063770993.

**Comment by kshitij12345** [comment]
I am seeing a test failure due to numerical mismatch for `test_te_linear_forward_backward_multiple_iteration` on v1.6 (works fine with v1.5). Will have a look. @crcrpar are you seeing this failure as well?

```python
>       assert_close(w1, te_linear1.weight)
E       AssertionError: Tensor-likes are not close!
E       
E       Mismatched elements: 16777216 / 16777216 (100.0%)
E       Greatest absolute difference: 9.361921547095688e+27 at index (0, 0) (up to 1e-05 allowed)
E       Greatest relative difference: 1.0078125 at index (0, 0) (up to 0.016 allowed)

thunder/tests/test_transformer_engine_executor.py:142: AssertionError
=================================================== short test summary info ====================================================
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[None] - AssertionError: Tensor-likes are not close!
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[fp8_recipe1] - AssertionError: Tensor-likes are not close!
========================================== 2 failed, 16 passed, 15 warnings in 6.18s ===========================================
```

File: thunder/executors/transformer_engineex.py [hunk 1]
```diff
@@ -52,6 +53,7 @@
         TE_AVAILABLE = False
 
     TE_VERSION_1_3_PLUS = LooseVersion(version("transformer_engine")) >= LooseVersion("1.3")
+    TE_VERSION_1_6_PLUS = LooseVersion(version("transformer_engine")) > LooseVersion("1.6")
```
Comment (nWEIdia) [inline thunder/executors/transformer_engineex.py:56]: Would transformer engine show up somewhere in requirements/ folder? Or is it an optional dependency?

Comment (IvanYashchuk) [reply thunder/executors/transformer_engineex.py:56]: It's an optional dependency and all tests should be skipped if TransformerEngine is not installed. 
There's a PR that adds TransformerEngine to be tested in CI https://github.com/Lightning-AI/lightning-thunder/pull/209#issuecomment-2074407065

**Comment by crcrpar** [comment]
> I am seeing a test failure due to numerical mismatch for `test_te_linear_forward_backward_multiple_iteration` on v1.6 (works fine with v1.5). Will have a look. @crcrpar are you seeing this failure as well?
> 
> ```python
> >       assert_close(w1, te_linear1.weight)
> E       AssertionError: Tensor-likes are not close!
> E       
> E       Mismatched elements: 16777216 / 16777216 (100.0%)
> E       Greatest absolute difference: 9.361921547095688e+27 at index (0, 0) (up to 1e-05 allowed)
> E       Greatest relative difference: 1.0078125 at index (0, 0) (up to 0.016 allowed)
> 
> thunder/tests/test_transformer_engine_executor.py:142: AssertionError
> =================================================== short test summary info ====================================================
> FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[None] - AssertionError: Tensor-likes are not close!
> FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[fp8_recipe1] - AssertionError: Tensor-likes are not close!
> ========================================== 2 failed, 16 passed, 15 warnings in 6.18s ===========================================
> ```

oh I'm seeing the same

**Comment by kshitij12345** [comment]
**The reason for failure**:

PR https://github.com/Lightning-AI/lightning-thunder/pull/48 introduced wrapping the generated trace in `fp8_autocast` context manager. Thing to note is that we apply this context manager in `no_grad` context.

```python
@no_grad
@fp8_autocast()
def computation_trace(*args, **kwargs):
```

However, this leads to problem in TE v1.6, because in this release, updating of forward pass scales and inverse scales  (and moving current amax to next position in buffer) is guarded by `torch.is_grad_enabled` which is disabled based on the order of decorators currently. Because of this, we don't actually update this meta-data correctly leading to incorrect output.

Ref: https://github.com/NVIDIA/TransformerEngine/blob/d07ef3da675d6fb594f4495899cc65b8ad477f68/transformer_engine/pytorch/fp8.py#L421-L422

-----
**Why it works in v1.5/stable**

This doesn't fail in v1.5 because this check is not guarded on `torch.is_grad_enabled` but instead on `module.is_training`

https://github.com/NVIDIA/TransformerEngine/blob/6a9edc38bf9b941b7d369af5103fa8fe0b121d61/transformer_engine/pytorch/module/base.py#L571

https://github.com/NVIDIA/TransformerEngine/blob/6a9edc38bf9b941b7d369af5103fa8fe0b121d61/transformer_engine/pytorch/module/base.py#L538-L550

-----

Potential Fixes:
1. Reorder the generated trace to apply `fp8_autocast` before `torch.no_grad`.
2. Remove automatic wrapping of `fp8_autocast` and let the user run the jitted function under this context (as before). They would have to apply it over the jitted function which would mean that unless they have disabled the grad (it should work).


cc: @IvanYashchuk

**Comment by IvanYashchuk** [comment]
> updating of forward pass scales and inverse scales (and moving current amax to next position in buffer) is guarded by `torch.is_grad_enabled` which is disabled based on the order of decorators currently.

Can we somehow wrap only the exit of the fp8 context manager in `torch.enable_grad`? Would changing the order of `@no_grad` and `@fp8_autocast()` work? If yes - let's do this.

For the future, maybe we need to create our own Thunder+TE context manager replicating https://github.com/NVIDIA/TransformerEngine/blob/d07ef3da675d6fb594f4495899cc65b8ad477f68/transformer_engine/pytorch/fp8.py#L558-L568
What do you both (Kshiteej and Masaki) think about this?

**Review by kshitij12345 — APPROVED** [review]
LGTM!

**Comment by kshitij12345** [comment]
@carmocca can you please have a look and stamp this, thanks!

**Review by carmocca — APPROVED** [review]
LGTM

**Comment by nWEIdia** [comment]
[celebrate]     Wei Wang reacted to your message:
________________________________

**Pull request merged**